_posts/2020-05-18-websocket-fallback-sockjs.md

```diff
@@ -0,0 +1,329 @@
+Problemy z RollUp zostały rozwiązane poprzez załadowanie biblioteki SockJS już po stronie klienta.
+
+
+### Wersja po uwzględnieniu poprawek
+#### Backend
+pom.xml
+{% highlight diff %}
+<dependencies>
+    <dependency>
+        <groupId>org.springframework.boot</groupId>
+        <artifactId>spring-boot-starter-websocket</artifactId>
+    </dependency>
++   <dependency> <!-- własna wersja sockjs jest opcjonalna :) -->
++       <groupId>org.webjars.npm</groupId>
++       <artifactId>sockjs-client</artifactId>
++       <version>1.4.0</version>
++   </dependency>
+</dependencies>
+{% endhighlight %}
+
+WebSocketConfig.java
+{% highlight diff %}
+@Configuration
+@EnableWebSocket
++@EnableWebMvc
+-public class WebSocketConfig implements WebSocketConfigurer {
++public class WebSocketConfig implements WebSocketConfigurer, WebMvcConfigurer {
+    private SocketHandler socketHandler;
+
+    public WebSocketConfig(SocketHandler socketHandler) {
+        this.socketHandler = socketHandler;
+    }
+
+    @Override
+    public void registerWebSocketHandlers(WebSocketHandlerRegistry registry) {
+-       registry.addHandler(socketHandler, "/signaling");
++       registry.addHandler(socketHandler, "/signaling"); // pozostawione dla kompatybilności wstecz :)
++       registry.addHandler(socketHandler, "/v2/signaling/sockjs-server") // endpoint dla SockJS
++               .withSockJS()
++               // SockJS wykorzystuje iframe w specyficznych przypadkach
++               // i tutaj "nadmiarowo" podajemy własną wersję biblioteki JS z własnego serwera,
++               // zamiast https://cdn.jsdelivr.net/sockjs/1.0.0/sockjs.min.js
++               .setClientLibraryUrl("/v2/signaling/sockjs-static/sockjs-client/sockjs.min.js");
+    }
++   @Override
++   public void addResourceHandlers(ResourceHandlerRegistry registry) {
++       // tutaj również nadmiarowo wystawiamy statyczny zasób z biblioteką SockJS
++       registry.addResourceHandler("/v2/signaling/sockjs-static/sockjs-client/*.min.js")
++               .addResourceLocations("classpath:/META-INF/resources/webjars/sockjs-client/1.4.0/dist/")
++               .setCachePeriod(3600)
++               .resourceChain(true)
++               .addResolver(new PathResourceResolver());
++    }
+}
+{% endhighlight %}
+
+#### Frontend
+web-socket-manager.ts
+{% highlight diff %}
+import {SockjsScriptLoader} from 'websocket/sockjs-script-loader';
+ 
+export class WebSocketManager {
+   private webSocket?: WebSocket;
+   constructor() {
+-       this.webSocket = this.initWebSocket();
++       new SockjsScriptLoader().loadSockJs()
++               .then(() => {
++                   this.webSocket = this.initWebSocket();
++               })
++               .catch(error => {
++                   this.webSocket = this.initWebSocket();
++               });
+   }
+
+    initWebSocket(): WebSocket {
++       // @ts-ignore
++       if (typeof SockJS === 'function') { // objeście problemu z RollUp ;)
```

Comment:
o rollupie już pisałem, wydaje się nadmiarowa taka informacja we wpisie o sockjs